### PR TITLE
WIP: Introduce PacketBatch with Growable buffer

### PR DESCRIPTION
**Pull request by steviez**
WIP: Still pulling stuff in, so this PR should be seeing some heavy adds for the next couple days. In any case, I'm being intentional with splitting things up per-commit. The "scaffolding" changes are currently left for sake of reducing the diff, and I have already identified a piece or two that can be carved out in their own PR's. Working on that to keep what will be the end diff reasonable.

#### Problem
We want to support transactions that exceed the current 1232 byte limit. In order to do so, we need to use a container that is capable of holding transactions that are larger in serialized size. Another PR (https://github.com/solana-labs/solana/pull/29055) took a blanket approach of adjusting the buffer for all Packets along the Tx processing path, but that approach uses additional memory for all transactions, even those that fit well within the original 1232 byte buffer.

#### Summary of Changes
Introduce a data structure, TxPacketBatch (leaning towards rename to `GrowablePacketBatch`). This data structure is meant to be a drop-in for the existing `PacketBatch` with following callouts:
- The buffer length is the same for all packets in the batch, but can be grown at runtime (ie if a caller wants more space than the default 1232 bytes)
- The buffers are laid out contiguously in memory (as opposed to a `Vec<Vec<u8>>`) for sake of memory locality (perf + ability to sigverify with GPU)

Fixes #

### CODE REVIEW

File: streamer/src/streamer.rs [hunk 1]
```diff
@@ -184,6 +189,92 @@ pub fn receiver(
...
+                    } = stats;
+
+                    packets_count.fetch_add(len, Ordering::Relaxed);
+                    packet_batches_count.fetch_add(1, Ordering::Relaxed);
+                    max_channel_len.fetch_max(packet_batch_sender.len(), Ordering::Relaxed);
+                    if len == PACKETS_PER_BATCH {
+                        full_packet_batches_count.fetch_add(1, Ordering::Relaxed);
+                    }
+
+                    let packet_batch = TxPacketBatch::from_packet_batch(packet_batch);
```
Comment (steviez) [inline streamer/src/streamer.rs:239]: This while `recv_loop2() & receiver2()` is a kind of a hack I made at the moment in lieu of deferring a design decision to review.

The question is whether we want to use the new data structure `TxPacketBatch` all throughout the codebase (Tx's, shreds, gossip), or keep its' use limited to the Tx path. As I have it coded now, I keep the change limited to the Tx processing path, and this is a shim for minimizing code duplication (can probably get something cleaner in the end result).

So, the direction question that I'd like feedback on is: Should we keep the existing PacketBatch data structure around for uses where we will continue to use UDP (and thus keep the 1232 byte restriction), or do we want to go all in and use the new data structure everywhere?

File: core/src/sigverify_stage.rs [hunk 2]
```diff
@@ -216,27 +202,11 @@ impl SigVerifierStats {
     }
 }
 
-impl SigVerifier for DisabledSigVerifier {
```
Comment (ilya-bobyr) [inline core/src/sigverify_stage.rs:219]: Maybe add a note into the commit description as to why you are removing this trait?
Looks like it had only one use case - part of the API of `TransactionSigVerifier`.
Was it an unnecessary abstraction?

File: perf/src/deduper.rs [hunk 3]
```diff
@@ -85,14 +83,14 @@ impl Deduper {
 
     pub fn dedup_packets_and_count_discards(
         &self,
-        batches: &mut [PacketBatch],
-        mut process_received_packet: impl FnMut(&mut Packet, bool, bool),
+        batches: &mut Vec<TxPacketBatch>,
```
Comment (ilya-bobyr) [inline perf/src/deduper.rs:86]: A change from `&mut [T]` to `&mut Vec[T]` is reducing generalization.
As the rest of the changes in this patch do not need that, maybe keep `&mut [T]`?

**Review by ilya-bobyr — CHANGES_REQUESTED** [review]

File: perf/src/tx_packet_batch.rs [hunk 4]
```diff
@@ -0,0 +1,697 @@
...
+    /// Shared buffer for all packets in batch
+    // TODO: change to u64 so we get guaranteed same alignment as Packet (8)
+    data: PinnedVec<u8>,
+    /// Current capacity of batch (packets)
+    capacity: usize,
+    /// Current length of batch (packets)
+    len: usize,
+    /// Length of batch packet's payload buffer (bytes)
+    buffer_len: usize,
+    /// Offset to ensure Meta references from buffer are properly aligned
```
[the remark this answers is not in the document]
Comment (steviez) [reply perf/src/tx_packet_batch.rs:45]: Fair, although this is going away per the other comment:
```
// TODO: change to u64 so we get guaranteed same alignment as Packet (8)
```
Basically,
- We have a Vector of u8's, u8's have alignment of 1
- Our buffer will "contain" Meta's that we'll reference out by byte address.
- The Meta's have an alignment of 8 (they contain a u64)
- So, this `alignment_offset` was added to provide a shift should the address to the start of the `Vec`'s contents falls on an odd address

Switching over to `Vec<u64>` will give the buffer an alignment of `8` as well, so no need for this alignment offset.

[the remark this answers is not in the document]
Comment (steviez) [reply perf/src/tx_packet_batch.rs]: I have made the above change (buffer of u64's), so alignment offset is no longer needed and has been removed.

File: core/src/sigverify_stage.rs (hunk 2 above)
Comment (steviez) [reply core/src/sigverify_stage.rs:219]: > Was it an unnecessary abstraction?

Yes, made things a little easier with this change. I intentionally have that change broken out into its' own commit, and should we proceed with this PR, I will open a separate PR for just that commit and rebase this PR on top

File: perf/src/tx_packet_batch.rs [hunk 5]
```diff
@@ -0,0 +1,697 @@
...
+
+        let data_len = Self::calc_data_len(capacity, buffer_len);
+        let mut data = PinnedVec::<u8>::with_capacity(data_len);
+        unsafe {
+            // We know data has data_len capacity, so set_len() with data_len
+            // is also safe. data is intentionally a raw buffer and it is up to
+            // the caller to do any initialization on contents of data.
+            data.set_len(data_len);
+        }
+        let alignment_offset = Self::calc_alignment_offset(&data);
```
[the remark this answers is not in the document]
Comment (steviez) [reply perf/src/tx_packet_batch.rs]: > Maybe just not setting the length would also be good enough?

Good points all around. I shifted the `set_len()` from `with_capacity()` to `with_size()`

File: perf/src/tx_packet_batch.rs [hunk 6]
```diff
@@ -0,0 +1,697 @@
...
+        unsafe { self.as_mut_ptr().add(index * self.buffer_len) }
+    }
+
+    #[inline]
+    /// Returns the number of packets the batch can hold without reallocating.
+    pub fn capacity(&self) -> usize {
+        self.capacity
+    }
+
+    #[inline]
```
[the remark this answers is not in the document]
Comment (steviez) [reply perf/src/tx_packet_batch.rs:346]: Fair, I definitely went overboard here. I stripped all of the `#[inline]` attributes except for the most basic functions (ie casting a pointer or return a member of the struct). This is now more inline with std library, such as [Vec](https://doc.rust-lang.org/src/alloc/vec/mod.rs.html)

File: perf/src/tx_packet_batch.rs [hunk 7]
```diff
@@ -0,0 +1,717 @@
...
+
+        // Cast pointer to u8 first as the calculated offset will be in bytes.
+        let offset_in_bytes = packet_index * self.buffer_len;
+        unsafe { self.as_ptr().cast::<u8>().add(offset_in_bytes) }
+    }
+
+    #[inline]
+    /// Returns a mutable pointer to the beginning of the data buffer.
+    pub fn as_mut_ptr(&mut self) -> *mut u8 {
+        unsafe { self.data.as_mut_ptr().cast::<u8>().add(0) }
```
Comment (steviez) [inline perf/src/tx_packet_batch.rs:328]: Oops, I'll clean this up

**Comment by steviez** [comment]
Had a good conversation with @ilya-bobyr - this approach temporarily on hold in favor of another

**Pull request closed without merging**